build/pack.c

```diff
@@ -525,20 +530,28 @@ static rpmRC writeRPM(Package pkg, unsigned char ** pkgidp,
     if (cpio_doio(fd, pkg, rpmio_flags, pld_algo, &archiveSize, &upld))
 	goto exit;
     payloadEnd = Ftell(fd);
+    payloadSize = payloadEnd - payloadStart;
 
     /* Re-read payload to calculate compressed digest */
     fdInitDigestID(fd, pld_algo, RPMTAG_PAYLOADDIGEST, 0);
-    if (fdConsume(fd, payloadStart, payloadEnd - payloadStart))
+    if (fdConsume(fd, payloadStart, payloadSize))
 	goto exit;
     fdFiniDigest(fd, RPMTAG_PAYLOADDIGEST, (void **)&pld, NULL, 1);
 
-    /* Insert the payload digests in main header */
+    /* Insert the payload digests + size in main header */
     headerDel(pkg->header, RPMTAG_PAYLOADDIGEST);
     headerPutString(pkg->header, RPMTAG_PAYLOADDIGEST, pld);
     headerDel(pkg->header, RPMTAG_PAYLOADDIGESTALT);
     headerPutString(pkg->header, RPMTAG_PAYLOADDIGESTALT, upld);
     pld = _free(pld);
 
+    if (pkg->rpmver >= 6) {
+	headerDel(pkg->header, RPMTAG_PAYLOADSIZE);
+	headerPutUint64(pkg->header, RPMTAG_PAYLOADSIZE, &payloadSize, 1);
+	headerDel(pkg->header, RPMTAG_PAYLOADSIZEALT);
+	headerPutUint64(pkg->header, RPMTAG_PAYLOADSIZEALT, &archiveSize, 1);
```

Comment:
I'm not fully sure I understand the use case for this

`RPMTAG_SIZE` / `RPMTAG_LONGSIZE` already covers the "how much space does this package require once installed" case.  

`RPMTAG_ARCHIVESIZE` / `RPMTAG_LONGARCHIVESIZE` already covers the uncompressed payload size (in theory it should be the same as `RPMTAG_PAYLOADSIZEALT`?), and it's *supposed* to be in the main header.  And I don't think the value overlaps?  Why introduce something new, except to make the name more consistent?

There's no equivalent for compressed payload size, but with either compressed *or* uncompressed payload size, you can't really usefully extract the freestanding archive from the RPM, because RPM is (now) the only tool that can process that archive.  It just seems unnecessary to even include that info.